content/blog/2024/testing-fastapi.md

```diff
@@ -0,0 +1,355 @@
+def engine():
+    """
+    Create the test database (if needed) and engine
+    """
+    test_db_name = "test_app"
+    db_url = urlparse(str(settings.DATABASE_URL))
+    test_engine = create_engine(db_url._replace(path=f"/{test_db_name}").geturl())
+
+    try:
+        Base.metadata.drop_all(bind=test_engine)
+    except OperationalError as err:
+        if f'database "{test_db_name}" does not exist' not in str(err):
+            raise
+        root_db_url = db_url._replace(path="/postgres").geturl()
+        conn = create_engine(root_db_url, isolation_level="AUTOCOMMIT").connect()
+        conn.execute(sa.text(f"CREATE DATABASE {test_db_name}"))
+        conn.close()
+    finally:
+        Base.metadata.create_all(bind=test_engine)
+
+    return test_engine
+```
+
+This [pytest fixture] is scoped to the test `"session"`, which means it will be
+automatically created once and shared across all tests. We start by obtaining a
+`db_url` from our application settings. This contains the credentials to connect
+to the database engine. We then create a new SQLAlchemy engine with the same
+credentials, but with a different database name defined in `test_db_name`.
+
+[pytest fixture]: https://docs.pytest.org/en/stable/explanation/fixtures.html
+
+We then `try` to drop the tables from the test database. If the database doesn't
+exist, we create it. Finally, we create all the tables in the test database. We
+now have an isolated database for our tests.
+
+Now let's review our application code to figure out how to insert this new
+engine:
+
+```python
+# File: app/main.py
+
+from fastapi import FastAPI, Depends
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+from app import settings
+
+app = FastAPI()
+
+engine = create_engine(settings.DATABASE_URL)
+
+SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
+
+def get_db():
+    db = SessionLocal()
+    try:
+        yield db
+    finally:
+        db.close()
+
+
+@app.get("/")
+def my_endpoint(db: Session = Depends(get_db)):
+    return db.query(...).all()
+```
+
+Notice the chain of dependencies starting from the `my_endpoint` function:
+
+1. `my_endpoint` uses a `db` parameter provided by `get_db` (wrapped in `Depends`).
+2. `get_db` yields from `SessionLocal`.
+3. `SessionLocal` is bound to the `engine` we created.
+4. `engine` is created from the `DATABASE_URL` in our settings.
+
+We can try to be clever and monkey-patch the settings or the functions provided
+by SQLAlchemy, but FastAPI actually provides a better way to do this:
+[dependency overrides]. Basically, we can replace any function or class that
+uses `Depends` with a different implementation.
+
+[dependency overrides]:
+    https://fastapi.tiangolo.com/advanced/testing-dependencies/
+
+Let's create a new fixture that will automatically inject itself in place of the
+`get_db` function:
+
+```python
+# File: conftest.py
+
+from app.main import app, get_db
+
+@pytest.fixture(autouse=True)
```

Comment:
This isn't a Session fixture, so it runs once for every test, correct? I think it would be worth pointing that out- it wasn't clear to me initially why the database was set up once in `engine` but reset everytime here.